### PR TITLE
Limelight pose estimation fixes and now recognizes apriltag and resets odometry according to how it is viewing apriltag

### PR DESCRIPTION
**Pull request by grace-mills**

### CODE REVIEW

File: Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java [hunk 1]
```diff
@@ -212,6 +227,14 @@ private void updateSmartDashboard( )
...
     SmartDashboard.putNumber("poseEstimationRotation", m_poseEstimator.getEstimatedPosition( ).getRotation( ).getDegrees( ));
 
+    Pose2d rawPoseSD = m_vision.getLimelightRawPose( );
+
+    if (rawPoseSD != null)
+    {
+      SmartDashboard.putNumber("rawPoseX", rawPoseSD.getX( ));
+      SmartDashboard.putNumber("rawPoseY", rawPoseSD.getY( ));
+    }
+
```
Comment (hpark-24) [inline Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java]: I was just wondering, if you are logging the "rawPose" do you also need to log "rawPoseX" and "rawPoseY" or are they different? If they are the same, maybe we should just keep one or the other (or, since "poseEstimation" also has the same values, just keeping "poseEstimation" and getting rid of the "rawPose" logs).

File: Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java [hunk 2]
```diff
@@ -503,6 +526,10 @@ public Pose2d getPose( )
 
   public void resetOdometry(Pose2d pose)
   {
+    DataLogManager.log("heading " + m_heading);
+    DataLogManager.log("positions " + getPositions( ));
+    DataLogManager.log("pose " + pose);
+
```
Comment (hpark-24) [inline Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java:511]: I think we can delete this since it was used for testing and will be putting logs every time "resetOdometry" is called (so very often).

**Review by hpark-24 — COMMENTED** [review]
It might be good to get rid of a couple logs for Main before we merge this in, but otherwise, it looks good!

File: Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java (hunk 1 above)
Comment (grace-mills) [reply Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java]: fixed

File: Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java (hunk 2 above)
Comment (grace-mills) [reply Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java:511]: I deleted my DataLogManager command for the raw pose but these are the ones we used before

File: Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java [hunk 3]
```diff
@@ -125,7 +125,6 @@ public Swerve(Vision vision)
 
     resetAnglesToAbsolute( );
     resetGyro(180.0); // All starting positions facing driver in field relative
-    resetOdometry(new Pose2d(0, 0, m_heading));
```
Comment (jmullins16) [inline Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java:128]: Not sure why this is deleted in the Comp2023-imported project. Seems like it was OK when called in the constructor.

File: Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java [hunk 4]
```diff
@@ -144,7 +143,20 @@ public void periodic( )
   @Override
   public void simulationPeriodic( )
   {
-    // This method will be called once per scheduler run when in simulation
+    // if (m_allowPoseEstimate)
```
Comment (jmullins16) [inline Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java]: Not sure why this was added to Compe2023-imported either. If not needed, just revert this change.

File: Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java (hunk 2 above)
Comment (jmullins16) [reply Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java:511]: resetOdometry shouldn't be called very often (f we're doing things right), but those log messages probably won't be needed in the long term. Ok to leave in for now if it helps.

File: Comp2024/src/main/java/frc/robot/Constants.java [hunk 5]
```diff
@@ -372,25 +372,83 @@ public enum ClimberMode
...
+        new Pose2d(new Translation2d(11.904726, 3.713226), new Rotation2d(Units.degreesToRadians(300))),    // AprilTag ID: 11
+        new Pose2d(new Translation2d(11.904726, 4.49834), new Rotation2d(Units.degreesToRadians(60))),      // AprilTag ID: 12
+        new Pose2d(new Translation2d(11.220196, 4.105148), new Rotation2d(Units.degreesToRadians(180))),    // AprilTag ID: 13
+        new Pose2d(new Translation2d(5.320792, 4.105148), new Rotation2d(Units.degreesToRadians(0))),       // AprilTag ID: 14
+        new Pose2d(new Translation2d(4.641342, 4.49834), new Rotation2d(Units.degreesToRadians(120))),      // AprilTag ID: 15
+        new Pose2d(new Translation2d(4.641342, 3.713226), new Rotation2d(Units.degreesToRadians(240)))      // AprilTag ID: 16
     ));
+
+    // Direction of goal relative to AprilTag 
+    public enum VIGoalDirection
```
Comment (jmullins16) [inline Comp2024/src/main/java/frc/robot/Constants.java:413]: These enums go with last year's game (driving a path left, center, or right of a grid). Should be deleted.

File: Comp2024/src/main/java/frc/robot/Constants.java [hunk 6]
```diff
@@ -372,25 +372,83 @@ public enum ClimberMode
...
+     * from limelight to front robot edge
+     */
+  }
+
+  /////////////////////////////////////////////////////////////////////////////
+  // Limelight driving alignment
+  /////////////////////////////////////////////////////////////////////////////
+  public static final class LLConsts
+  {
+    // Default calibration
```
Comment (jmullins16) [inline Comp2024/src/main/java/frc/robot/Constants.java:432]: These constants are all a part of limelight driving algorithm used on Festus (and kept around for Slink, but never used). Should be delted.

File: Comp2024/src/main/java/frc/robot/subsystems/Vision.java [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+ *
+ */
+public class Vision extends SubsystemBase
+{
+  // Objects
+  public MedianFilter           m_tyfilter      = new MedianFilter(5); // median filter y values to remove outliers (5 sample)
+  public MedianFilter           m_tvfilter      = new MedianFilter(5); // median filter v values to remove outliers (5 sample)
+
+  // Declare module variables
+  private double                m_distance1     = LLConsts.kDistance1;   // x position in inches for first reference point
```
Comment (jmullins16) [inline Comp2024/src/main/java/frc/robot/subsystems/Vision.java]: This group of variables is also about LL driving algorithm and should be deleted. (down to the constructor)

File: Comp2024/src/main/java/frc/robot/subsystems/Vision.java [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+  public Vision( )
+  {
+    setName("Vision");
+    setSubsystem("Vision");
+
+    // Get the Network table reference once for all methods
+    m_table = NetworkTableInstance.getDefault( ).getTable("limelight");
+
+    // Put all the needed widgets on the dashboard
+    SmartDashboard.putNumber("VI_distance1", m_distance1);
```
Comment (jmullins16) [inline Comp2024/src/main/java/frc/robot/subsystems/Vision.java]: This group is from LL driving algorithm...

File: Comp2024/src/main/java/frc/robot/subsystems/Vision.java [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+    SmartDashboard.putNumber("VI_distance2", m_distance2);
+    SmartDashboard.putNumber("VI_vertOffset1", m_vertOffset1);
+    SmartDashboard.putNumber("VI_vertOffset2", m_vertOffset2);
+
+    SmartDashboard.setDefaultBoolean("VI_OVERRIDE", false);
```
Comment (jmullins16) [inline Comp2024/src/main/java/frc/robot/subsystems/Vision.java:64]: This group is from LL driving simulation...

File: Comp2024/src/main/java/frc/robot/subsystems/Vision.java [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+    SmartDashboard.putNumber("VI_OVERRIDE_TX", 0.0);
+    SmartDashboard.putNumber("VI_OVERRIDE_TY", 0.0);
+    SmartDashboard.putNumber("VI_OVERRIDE_TA", 0.0);
+    SmartDashboard.putNumber("VI_OVERRIDE_TS", 0.0);
+    SmartDashboard.putNumber("VI_OVERRIDE_TV", 0.0);
+
+    SmartDashboard.putNumberArray("VI_RobotPose", new double[ ] { });
```
Comment (jmullins16) [inline Comp2024/src/main/java/frc/robot/subsystems/Vision.java:71]: This should be kept (pose, area, skew, valid, latency), since we'll use them.

**Review by jmullins16 — CHANGES_REQUESTED** [review]
Lot's of great stuff in here and headed in the right direction. Please remove the files "touched" in Comp2023-imported from the PR. Also, it will probably help if we walk through the vision file, so I can help cut out stuff associated with older games/algorithms.

File: Comp2023-Imported/src/main/java/frc/robot/Constants.java [hunk 11]
```diff
@@ -512,19 +512,13 @@ public enum VIRequests
...
-      DIRECTION_RIGHT   // Right
-    }
 
-    public static final double kATagDepthInGrid    = Units.inchesToMeters(14.25);    // Depth from front of grid to AprilTag - 1'2-1/4"
-    public static final double kRobotCenterToFront = Units.inchesToMeters((28.0 + 6.0) / 2); // Depth from limelight to front robot edge
-    public static final double kAdjustPathX        = kATagDepthInGrid + kRobotCenterToFront;
-    public static final double kAdjustPathY        = Units.inchesToMeters(18.25 / 2 + 18.5 / 2) + 0.06;     // Addition of 6cm to adjust for empirical error 
-    public static final double kAdjustSubPathX     = kRobotCenterToFront + Units.inchesToMeters(30); // Robot stop 30 inches from the substation loading zone
-    public static final double kAdjustSubPathY     = Units.inchesToMeters(50.5 / 2);
+    public static final double       kATagDepthInGrid    = Units.inchesToMeters(14.25);    // Depth from front of grid to AprilTag - 1'2-1/4"
```
Comment (jmullins16) [inline Comp2023-Imported/src/main/java/frc/robot/Constants.java]: These are all geometry for last year's game. Note the reference to the grid. And the path adjustments were to the nodes with respect to the center of the robot.

File: Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java [hunk 12]
```diff
@@ -241,89 +228,18 @@ private void updateSmartDashboard( )
...
-    // load in Pid constants to controller
-    m_turnPid = new PIDController(m_turnPidKp, m_turnPidKi, m_turnPidKd);
-    m_throttlePid = new PIDController(m_throttlePidKp, m_throttlePidKi, m_throttlePidKd);
 
     vision.m_tyfilter.reset( );
     vision.m_tvfilter.reset( );
     vision.syncStateFromDashboard( );
   }
 
-  public void driveWithLimelightExecute(Vision vision)
```
Comment (jmullins16) [inline Comp2023-Imported/src/main/java/frc/robot/subsystems/Swerve.java]: If you're leaving in some commented code for examples, probably want to at least leave in an empty driveWithLimelightExecute{} method which will be called during the command.

File: Comp2023-Imported/src/main/java/frc/robot/subsystems/Vision.java [hunk 13]
```diff
@@ -29,17 +28,12 @@ public class Vision extends SubsystemBase
   public MedianFilter           m_tvfilter      = new MedianFilter(5); // median filter v values to remove outliers (5 sample)
 
   // Declare module variables
-  private double                m_distance1     = LLConsts.kDistance1;   // x position in inches for first reference point
-  private double                m_vertOffset1   = LLConsts.kVertOffset1; // y reading in degrees for first reference point
-  private double                m_distance2     = LLConsts.kDistance2;   // x position in inches for second reference point
-  private double                m_vertOffset2   = LLConsts.kVertOffset2; // y reading in degrees for second reference point
   private double                m_slope;   // Linear regressions slope from calibration
```
Comment (jmullins16) [inline Comp2023-Imported/src/main/java/frc/robot/subsystems/Vision.java]: slope and offset are also part of the festus limelight driving calibration--so shouldn't be needed.

**Review by jmullins16 — CHANGES_REQUESTED** [review]
I think I got a little confused because the changes spanned the two projects. If possible, just have all changes in the Comp2024 project, and revert those in Comp2023-imported. Overall, the changes look good--just a little confusing because they look like duplication.

**Review by jmullins16 — APPROVED** [review]
Looks good. Approving and merging. We can always continue to fix small things as we find them.
